src/script/microbit-interfacing/microbit-bluetooth.ts

```diff
@@ -0,0 +1,500 @@
+  }
+  for (const [key, value] of ioPinMessages) {
+    sendIOPinMessage(outputIO, { pin: key, on: value !== 0 });
+  }
+};
+
+const resetIOPins = (outputIO: BluetoothRemoteGATTCharacteristic) => {
+  StaticConfiguration.supportedPins.forEach(value => {
+    sendIOPinMessage(outputIO, { pin: value, on: false });
+  });
+};
+
+const subarray = <T>(arr: T[], start: number, end: number): T[] => {
+  const newArr: T[] = [];
+  for (let i = start; i < end; i++) {
+    newArr.push(arr[i]);
+  }
+  return newArr;
+};
+
+const setOutputMatrix = (
+  outputMatrix: BluetoothRemoteGATTCharacteristic,
+  matrix: boolean[],
+): void => {
+  const dataView = new DataView(new ArrayBuffer(5));
+  for (let i = 0; i < 5; i++) {
+    dataView.setUint8(
+      i,
+      subarray(matrix, i * 5, 5 + i * 5).reduce(
+        (byte, bool) => (byte << 1) | (bool ? 1 : 0),
+        0,
+      ),
+    );
+  }
+  addToServiceActionQueue(outputMatrix, dataView);
+};
+
+/**
+ * Sends a message through UART
+ * @param type The type of UART message, i.e 'g' for gesture and 's' for sound
+ * @param value The message
+ */
+const sendToOutputUart = (
+  outputUart: BluetoothRemoteGATTCharacteristic,
+  type: UARTMessageType,
+  value: string,
+): void => {
+  const view = MBSpecs.Utility.messageToDataview(`${type}_${value}`);
+  addToServiceActionQueue(outputUart, view);
+};
+
+const addToServiceActionQueue = (
+  service: BluetoothRemoteGATTCharacteristic,
+  view: DataView,
+) => {
+  bluetoothServiceActionQueue.update(update => {
+    update.queue.push({ service, view });
+    return update;
+  });
+  processServiceActionQueue();
+};
+
+const processServiceActionQueue = () => {
+  if (
+    get(bluetoothServiceActionQueue).busy ||
+    get(bluetoothServiceActionQueue).queue.length == 0
+  )
+    return;
+  get(bluetoothServiceActionQueue).busy = true;
+  const { service, view } = get(bluetoothServiceActionQueue).queue.shift() ?? {
+    service: undefined,
+    view: undefined,
+  };
+  if (service === undefined) {
+    throw new Error(
+      'Could not process the service queue, an element in the queue was not provided with a service to execute on.',
+    );
+  }
+
+  service
+    .writeValue(view)
+    .then(() => {
+      get(bluetoothServiceActionQueue).busy = false;
+      processServiceActionQueue();
+    })
+    .catch(err => {
+      // Catches a characteristic not found error, preventing further output.
+      // Why does this happens is not clear
+      console.error(err);
+      if (err) {
+        if ((err as DOMException).message.includes('GATT Service no longer exists')) {
+          // TODO: This is a bit of a tangled mess.
+          // listenToOutputServices()
+          //   .then(() => {
+          //     console.log('Attempted to fix missing gatt!');
+          //   })
+          //   .catch(() => {
+          //     console.error(
+          //       'Failed to fix missing GATT service issue. Uncharted territory',
+          //     );
+          //   });
```

Comment:
This should probably fallback to safely disconnecting and reconnecting the output device. Not important for initial release as output is not being used.